Add writting and array modifying features to Lethe_pyvista_tools (#686)
lethe-cfd/lethe: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst

Features
----------------------------------
- DEM simulation
- Post-processing using `Python <https://www.python.org/>`_, `PyVista <https://docs.pyvista.org/>`_, `lethe_pyvista_tools <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_, and `ParaView <https://www.paraview.org/>`_.

Files used in this example
----------------------------

- Parameters file for particle insertion: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/packing-rotating-drum.prm``
- Parameters file for drum rotation: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/small-rotating-drum-dem.prm``
- Python module for Lethe data post-processing: ``/contrib/postprocessing/lethe_pyvista_tools.py``

> - Python script using module for rotating drum post-processing: ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py``

Reviewer: Why ``/examples/postprocessing/small-scale-rotating-drum-postprocessing/example_small_rotating_drum.py``?
Is not the example in the examples/dem repository?

Description of the case
-----------------------

In this example, we illustrate the mixing inside a rotating drum by coloring the particles at the first time-step according to their radial position. To do so, we post-process Lethe-DEM data using `Python <https://www.python.org/>`_, `PyVista <https://docs.pyvista.org/>`_, `lethe_pyvista_tools <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_, and `ParaView <https://www.paraview.org/>`_

The DEM files used in this example are obtained following the `Small scale rotating drum example <../../dem/rotating-drum/small-scale-rotating-drum.html>`_.

.. note::

>   It is not necessary to use all mentioned tools, but they are used in this example to show different possibilities of data access according to user's need.

Reviewer 2: ```suggestion
  It is not necessary to use all mentioned tools, but they are used in this example to show different ways to process the data according to user's need.
```

Reading Lethe results to Python
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Next, we read the results into Python:

.. code-block::
  
  particles.read_lethe_to_pyvista(pvd_name = "out.pvd")

``read_lethe_to_pyvista`` takes the name of the ``.pvd`` file generated by the simulation. The method can take 3 other arguments: ``first``, ``last``, and ``interval``, standing for the first and last time-steps to be read and the interval between the time-steps, respectively. By default, ``first = 0``, ``interval = 1``, and ``last`` is the last time-step of the data.

The ``read_lethe_to_pyvista`` reading function assigns the datasets of each time-step to the object ``particles``. Each time-step corresponds to a `PyVista dataset <https://docs.pyvista.org/user-guide/vtk_to_pyvista.html#>`_, and can be accessed using ``particles.df[$TIME-step_NUMBER]``.

.. note:: 
  
  If we set ``first = 2``, ``particles.df[0]`` corresponds to the dataset of the 2nd time-step of the simulation.

Creation of a new array
~~~~~~~~~~~~~~~~~~~~~~~~

To color the particles according to their position, we use the function ``array_modifier``, which takes the following arguments:

> - ``reference_array_name``: Name of the array used to sort the data. By default: ``"ID"``

Reviewer: I know that saying "array used to sort the data" and "ID" means you use the ID of particle for sorting, but what else can it be? If there's other nice option, how can it be relevant?

- ``array_name``: Name of the new array. If there is an array with the same name, it will be rewritten according to the other arguments. By default: ``"new_array"``
- ``restart_array``: If ``True``, zeroes the entire array before modifying it. If one wants to modify part of the array keeping the rest intact, it must be set as ``False``. By default: ``False``
- ``condition``: Takes a string and uses it in an if condition to modify the array. Variables accepted include ``x``, ``y``, ``z``, ``u``, ``v``, ``w``, ``t``, and any other array (``ID`` for example). It also accepts a combination of them, such as ``"x*w**2 + t > 2 and ID > 0"``. By default: ``""``
- ``array_values``: New values to the array. This argument accepts a single value (which will be repeated to all data respecting the given ``condition``), a `NumPy <https://numpy.org/>`_ array, or `Python list <https://docs.python.org/3/tutorial/datastructures.html>`_ (with the same length (``len``) of all other arrays in ``particles.df``), or a string such as ``"2*x + t"`` (working just like the ``condition`` argument). By default: ``0``

> - ``standard_value``: If ``restart array = True`` or the array is a completely new array, the ``standard_value`` will be plugged to the entire array before modifying it. By default: ``0``

Reviewer: I think the description of the line 105 (restart_array) should not says "zeroes the entire array before modifying it" if you can choose this value with this parameters, even if 0 is the default value :)

- ``time_dependent``: ``array_modifier`` can be time dependent or not. If set ``True``, ``condition`` will be tested to each of the time-steps, while if it is ``False``, it will be applied using the ``reference_time_step`` instead, and the modification will be just replicated to the other time-steps. By default: ``False``

> - ``reference_time_step``: Reference time-step to which the modification will be applied. If ``time_dependent = False``, the result of the modificaition applied to ``reference_time_step`` will be simply replicated to the others. By default: ``0``

Reviewer: This parameter is so powerful!! :D 
As already said in this PR, I think you should explain that this is not really the time step, but the output step.

.. code-block::
  
  particles.write_vtu(prefix = "mod_")

This will save all data in the output folder of the simulation. All files will have a ``prefix`` and can be opened in ParaView just like any other Lethe simulation.

.. image:: images/paraview_window.png
    :alt: ParaView
    :align: center

 

.. image:: images/paraview_window_zoom.png
    :alt: ParaView zoom in
    :align: center

Results
---------

Following we present one video of the full simulation with particles colored by their initial position.

.. raw:: html

  <iframe width="560" height="315" src="https://www.youtube.com/embed/qxO4MD_zg2w" title="Rotating drum - mixing study" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

The simulation results show that the layers in red and green, that is, particles close to the walls, tend to mix faster than the ones in the center.

Possibilities for extension
----------------------------

- Give a different ``condition`` to create the ``particle_color`` array
- Use the ``lethe_pyvista_tools`` for a different problem, modifying the ``condition`` accordingly

> - Use the tools in the `PyVista oficial repository <https://docs.pyvista.org>`_ to create screenshots, movies, and plots with the data.

Reviewer: ```suggestion
- Use the tools in the `PyVista official repository <https://docs.pyvista.org>`_ to create screenshots, movies, and plots with the data.
```
